auth.py

```diff
@@ -43,12 +59,20 @@ def callback():
         mappings = current_app.config["SAML"]["attributes"]
         attrs = {key: auth_attrs.get(mapping, []) for key, mapping in mappings.items()}
 
+        uid = attrs["uid"][0]
+        first_name = attrs["givenName"][0] if len(attrs["givenName"]) else ''
+        name = attrs["sn"][0]
+        email = attrs["email"][0]
+
         session["logged_in"] = True
-        session["user_id"] = attrs["uid"][0]
-        session["first_name"] = attrs["givenName"][0] if len(attrs["givenName"]) else ''
-        session["name"] = attrs["sn"][0]
-        session["email"] = attrs["email"][0]
+        session["uid"] = uid
+        session["first_name"] = first_name
+        session["name"] = name
+        session["email"] = email
 
+        user = User(name=name, first_name=first_name, email=email)
+        user_id = check_or_create_user(user)
```

Comment:
Here you will always set the session name to the directory value at runtime. People that log into the app will always see their name as the updated directory one. However, these changes are not saved in database.

If Michel Dupont changes their name to Michelle Dupont, Michelle will see their name updated, while the administrators or other users will still see them as Michel, leading to a consistency issue.

I'd recommend, either:
- Always update user names from the directory value and trust identity based on returned email address and disallow users to change their names manually.
- Set the session name to the one stored in database. All users will see a consistent name and will have to update their name from the app if their name changes.

Reply:
By the directory value, do you mean the value sent by the SAML idp ?